crates/ursa-rpc-service/src/api.rs

```diff
@@ -317,3 +456,27 @@ where
         Pin::new(&mut self.reader).poll_read(cx, buf)
     }
 }
+
+pub async fn load_car_checked<R, B>(s: &B, reader: R, root_cid: Cid) -> Result<Vec<Cid>>
```

Comment:
the latter half is still a copy of the `load_car`. we which prob call `load_car`

Reply:
how do you propose the former half be done then? without copying data

Reply:
I am trying to figure out if I follow. This the load car impl, https://github.com/filecoin-project/ref-fvm/blob/3baa50edf5c15d2645a7dc3cfbab718831397646/ipld/car/src/lib.rs#L155

and you just copy-paste this here:

```rust
    let mut buf = Vec::with_capacity(100);
    while let Some(block) = car_reader.next_block().await? {
        buf.push((block.cid, block.data));
        if buf.len() > 1000 {
            s.put_many_keyed(buf.iter().map(|(k, v)| (*k, v)))
                .map_err(|e| Error::Other(e.to_string()))?;
            buf.clear();
        }
    }
    s.put_many_keyed(buf.iter().map(|(k, v)| (*k, v)))
        .map_err(|e| Error::Other(e.to_string()))?;
    Ok(car_reader.header.roots)
```

Reply:
CarReader consumes the async reader to parse the headers, so I can't get the header and then pass the inner reader from CarReader to load_car, because it'll be missing headers and fail due to cbor. As a workaround this method performs the root cid check and the same logic from load car with a single CarReader, since nesting them would error. Of course the other solution would be to clone the data and parse twice but that's not nearly as efficient. 

I think a good improvement to the fvm_ipld_car crate would be to provide an implementation of load_car that when passed a CarReader it would correctly reuse headers and blocks

Reply:
we can open PR there

Reply:
Just reading over it now i don't this is a necessary step 

```
    // verify root cid is present in the car header
    if !car.header.roots.contains(&root_cid) {
        return Err(anyhow!("Root CID not present in car header"));
    }
```

Using the normal `load_car` vs `load_car_unchecked` should check the cids according to docs I linked above

Reply:
While the origin most likely respond with a valid CAR, i think there's a chance in a malicious scenario that the car could be incorrect, and we'd store data we don't care about. Do we want to 100% trust that origin gateways are serving car requested?

Reply:
There is not trust assumptions here. It's a car file

Reply:
https://ipld.io/specs/transport/car/

Another good reference.

Reply:
```
match cid.hash().code() {
    0x0 => {
        if cid.hash().digest() != data {
            return Err(Error::InvalidFile(
                "CAR has an identity CID that doesn't match the corresponding data"
                    .into(),
            ));
        }
    }
    code => {
        let code = multihash::Code::try_from(code)?;
        let actual = Cid::new_v1(cid.codec(), code.digest(&data));
        if actual != cid {
            return Err(Error::InvalidFile(format!(
                "CAR has an incorrect CID: expected {}, found {}",
                cid, actual,
            )));
        }
    }
}
```
This is the verification, It's only making sure the CAR is valid to itself, not to the request we made to the gateway. Invalid CAR would be rejected here, but we still should make sure that the valid car we get back is for the root cid we intended to receive specifically, and not some arbitrary valid CAR

Reply:
I need to follow how to check the root. You can get the root of the car file after you load it. This seems redundant. We can just go with using `load_car`

> The roots array must contain one or more CIDs, each of which should be present somewhere in the remainder of the CAR.

Reply:
If we get the root after it's been loaded and it's missing, should anything be done about the arbitrary content just loaded?

Reply:
https://github.com/fleek-network/ursa/pull/278/commits/dc750a096aeee288a693c195e2bbd34bd205227c